lib/promscrape/scrapework.go

```diff
@@ -527,51 +527,64 @@ func (sw *scrapeWork) processDataOneShot(scrapeTimestamp, realTimestamp int64, b
 }
 
 func (sw *scrapeWork) processDataInStreamMode(scrapeTimestamp, realTimestamp int64, body *bytesutil.ByteBuffer, scrapeDurationSeconds float64) error {
-	samplesScraped := 0
-	samplesPostRelabeling := 0
-	wc := writeRequestCtxPool.Get(sw.prevLabelsLen)
+	var samplesScraped atomic.Int64
+	var samplesPostRelabeling atomic.Int64
+	var samplesDroppedTotal atomic.Int64
 
 	lastScrape := sw.loadLastScrape()
 	bodyString := bytesutil.ToUnsafeString(body.B)
 	areIdenticalSeries := sw.areIdenticalSeries(lastScrape, bodyString)
-	samplesDropped := 0
 
 	r := body.NewReader()
-	var mu sync.Mutex
 	err := stream.Parse(r, scrapeTimestamp, "", false, func(rows []parser.Row) error {
-		mu.Lock()
-		defer mu.Unlock()
+		var err error
 
-		samplesScraped += len(rows)
+		// stream.Parse internally allocates at least a 64KB buffer.
+		// To minimize memory allocations and improve throughput, we preallocate
+		// a buffer of the same size.
+		//
+		// See https://github.com/VictoriaMetrics/VictoriaMetrics/blob/f645479b5e0612e7e781aa50a0375c3624b3adfe/lib/protoparser/protoparserutil/lines_reader.go#L42
+		wc := writeRequestCtxPool.Get(protoparserutil.DefaultBlockSize)
+		defer func() {
+			wc.resetNoRows()
```

Comment:
looks like wc.resetNoRows is already unconditionally called on L:576

Reply:
fixed